File: agent/agents/orchestrator.py
"""Orchestrator agent (PRD S6.2): decomposes a job, pays specialists via
x402, hands the assembled claims to the evaluator, then triggers
settlement. Owns the JobRecord end to end since it's the only role that
needs to persist job state across the whole lifecycle.

Run standalone with:
    python -m agents.orchestrator
This is also the one HTTP surface a frontend integrates against later:
POST /jobs to submit work, GET /jobs/{id} to poll status, GET /reputation
for the dashboard feed.
"""
from __future__ import annotations
import uuid
from datetime import datetime, timezone

import httpx
from fastapi import FastAPI, HTTPException
from fastapi.middleware.cors import CORSMiddleware

from shared.schema import JobRequest, JobRecord, Claim
from shared.console import log, rule
from shared.config import (
    ORCHESTRATOR_PORT, ONCHAIN_AGENT_URL, NEWS_AGENT_URL, COMPLIANCE_AGENT_URL,
    EVALUATOR_URL, WALLETS, ANTHROPIC_API_KEY, OPENAI_API_KEY, GOOGLE_API_KEY,
    ARC_CHAIN_ID, ARC_RPC_URL, ARC_EXPLORER_URL, NANOPAYMENT_USDC, FRONTEND_ORIGIN,
)
from payments.x402 import x402_post
from payments.wallet import ledger, transfer_to_address
from payments import chain
from storage.store import job_store, reputation_store
from settlement.escrow import settle
from agents import langchain_planner

app = FastAPI(title="VeriFi Orchestrator")

app.add_middleware(
    CORSMiddleware,
    allow_origins=[FRONTEND_ORIGIN] if FRONTEND_ORIGIN != "*" else ["*"],
    allow_credentials=False,
    allow_methods=["GET", "POST"],
    allow_headers=["*"],
)

ORCHESTRATOR_PRIVATE_KEY = WALLETS["orchestrator"]["private_key"]
ESCROW_ADDRESS = WALLETS["escrow"]["address"]

SPECIALIST_URLS: dict[str, str] = {
    "onchain-agent-v1": ONCHAIN_AGENT_URL,
    "news-agent-v1": NEWS_AGENT_URL,
    "compliance-agent-v1": COMPLIANCE_AGENT_URL,
}

# Fixed Template A/B subtask decomposition (PRD S8) -- used only as a
# fallback when no LLM is configured. When ANTHROPIC_API_KEY or
# OPENAI_API_KEY is set, `decompose()` instead asks a real LangChain
# tool-calling agent (agents/langchain_planner.py) which specialists this
# specific request actually needs.
TEMPLATE_SPECIALISTS: dict[str, list[str]] = {
    "protocol_treasury_diligence": ["onchain-agent-v1", "news-agent-v1", "compliance-agent-v1"],
    "yield_opportunity_review": ["onchain-agent-v1", "compliance-agent-v1"],
}


async def decompose(job_req: JobRequest) -> list[str]:
    if GOOGLE_API_KEY or ANTHROPIC_API_KEY or OPENAI_API_KEY:
        try:
            plan = await langchain_planner.plan_specialists(job_req)
            return plan.specialist_ids
        except Exception as e:
            # Free-tier LLM quotas (daily caps, not per-minute) are a real
            # external dependency this job shouldn't die on -- the payment/
            # verification pipeline underneath doesn't care who picked the
            # specialist list. Fall back to the fixed template rather than
            # failing (and having already locked the requester's budget).
            log("orchestrator", f"LLM planner unavailable ({e}) -- falling back to fixed Template A/B list", style="grey62")
            return TEMPLATE_SPECIALISTS[job_req.template]
    log("orchestrator", "no LLM configured -- using fixed Template A/B specialist list (see README)", style="grey62")
    return TEMPLATE_SPECIALISTS[job_req.template]


async def call_specialist(client: httpx.AsyncClient, name: str, job_req: JobRequest, job_id: str) -> list[Claim]:
    url = SPECIALIST_URLS[name]
    log("orchestrator", f"-> calling {name} ({url}/analyze)")
    resp = await x402_post(client, f"{url}/analyze", ORCHESTRATOR_PRIVATE_KEY, {
        "job_id": job_id,
        "protocol_slug": job_req.protocol_slug,
        "target_address": job_req.target_address,
        "inject_fault": job_req.inject_fault,
    })
    resp.raise_for_status()
    return [Claim(**c) for c in resp.json()["claims"]]


async def assemble_memo(job_req: JobRequest, claims: list[Claim]) -> str:
    if GOOGLE_API_KEY or ANTHROPIC_API_KEY or OPENAI_API_KEY:
        try:
            return await langchain_planner.write_memo(job_req, claims)
        except Exception as e:
            log("orchestrator", f"LLM memo writer unavailable ({e}) -- falling back to templated memo", style="grey62")
            return _template_memo(job_req, claims)
    return _template_memo(job_req, claims)


def _template_memo(job_req: JobRequest, claims: list[Claim]) -> str:
    lines = [f"# Diligence Memo: {job_req.protocol_slug}", "", job_req.request_text, ""]
    by_type: dict[str, list[Claim]] = {}
    for c in claims:
        by_type.setdefault(c.claim_type, []).append(c)
    for claim_type, group in by_type.items():
        lines.append(f"## {claim_type}")
        for c in group:
            lines.append(f"- {c.claim_text} (source: {c.provider_source})")
        lines.append("")
    return "\n".join(lines)


@app.post("/jobs")
async def create_job(job_req: JobRequest):
    job_id = f"job_{uuid.uuid4().hex[:10]}"
    rule(f"NEW JOB {job_id} -- {job_req.protocol_slug}")
    log("orchestrator", f'received request: "{job_req.request_text}" | budget={job_req.budget_usdc} USDC | template={job_req.template}')

    job = JobRecord(
        job_id=job_id,
        requester_id=job_req.requester_wallet,
        template=job_req.template,
        request_text=job_req.request_text,
        protocol_slug=job_req.protocol_slug,
        budget_usdc=job_req.budget_usdc,
        status="in_progress",
        created_at=datetime.now(timezone.utc).isoformat(),
    )

    if job_req.payment_tx_hash:
        # Real connected-wallet path: the requester already broadcast a real
        # transfer to escrow themselves. Don't trust the claim -- re-derive
        # it from the chain, same principle as payments/x402.py's
        # verify_transfer use for specialist nanopayments.
        verified = chain.verify_transfer(
            job_req.payment_tx_hash,
            expected_from=job_req.requester_wallet,
            expected_to=ESCROW_ADDRESS,
            min_amount_usdc=job_req.budget_usdc,
        )
        if not verified:
            job.status = "failed"
            job_store.save(job)
            raise HTTPException(status_code=402, detail=f"payment tx {job_req.payment_tx_hash} could not be verified on-chain")
        log("orchestrator", f"budget verified on-chain: {job_req.requester_wallet[:10]}.. -> escrow (tx {job_req.payment_tx_hash})")
    else:
        # Fallback path used by cli/run_demo.py and other backend-internal
        # callers that don't have a real connected wallet -- signs the
        # transfer with the fixed 'requester' role wallet from .env.
        try:
            ledger.transfer(job_req.requester_wallet, "escrow", job_req.budget_usdc, memo=f"job {job_id}: lock budget")
        except Exception as e:
            job.status = "failed"
            job_store.save(job)
            raise HTTPException(status_code=402, detail=f"could not lock budget in escrow: {e}")

    try:
        specialists = await decompose(job_req)
        job.subtasks = specialists

        all_claims: list[Claim] = []
        async with httpx.AsyncClient(timeout=30) as client:
            for name in specialists:
                try:
                    claims = await call_specialist(client, name, job_req, job_id)
                    all_claims.extend(claims)
                except Exception as e:
                    log("orchestrator", f"  ! {name} failed: {e}", style="bold red")

        job.claims = all_claims
        job.final_memo = await assemble_memo(job_req, all_claims)

        log("orchestrator", f"assembled {len(all_claims)} claims from {len(specialists)} specialists -> handing off to evaluator")
        async with httpx.AsyncClient(timeout=30) as client:
            eval_resp = await client.post(f"{EVALUATOR_URL}/evaluate", json={
                "job_id": job_id,
                "protocol_slug": job_req.protocol_slug,
                "target_address": job_req.target_address,
                "claims": [c.model_dump() for c in all_claims],
            })
            eval_resp.raise_for_status()
            job.claims = [Claim(**c) for c in eval_resp.json()["claims"]]

        job = await settle(job)
        job_store.save(job)

        log("orchestrator", f"job {job_id} DONE -- verdict={job.overall_verdict.upper()} paid={job.total_paid_usdc:.4f} USDC", style="bold cyan")
        return job.model_dump()

    except Exception as e:
        # Budget is already locked in escrow at this point (see above) --
        # any failure past that line must not leave real funds stranded
        # with no job record and no way back to the requester. Refund for
        # real, on-chain, same as any other settlement transfer.
        log("orchestrator", f"job {job_id} failed after budget lock: {e} -- refunding escrow -> requester", style="bold red")
        job.status = "failed"
        try:
            if job_req.payment_tx_hash:
                # Real connected wallet -- refund straight to that address,
                # not through the role lookup `ledger.transfer` uses.
                transfer_to_address("escrow", job_req.requester_wallet, job_req.budget_usdc, memo=f"job {job_id}: refund after failure")
            else:
                ledger.transfer("escrow", job_req.requester_wallet, job_req.budget_usdc, memo=f"job {job_id}: refund after failure")
        except Exception as refund_error:
            log("orchestrator", f"  ! refund ALSO failed: {refund_error} -- funds remain in escrow, job marked failed", style="bold red")
        job_store.save(job)
        raise HTTPException(status_code=500, detail=f"job failed after budget was locked; escrow refunded to requester where possible: {e}")


@app.get("/jobs/{job_id}")
async def get_job(job_id: str):
    job = job_store.get(job_id)
    if not job:
        raise HTTPException(status_code=404, detail="job not found")
    return job.model_dump()


@app.get("/jobs")
async def list_jobs():
    return [j.model_dump() for j in job_store.list_all()]


@app.get("/reputation")
async def get_reputation():
    return [r.model_dump() for r in reputation_store.list_all()]


@app.get("/wallets")
async def get_wallets():
    return ledger.all_balances()


@app.get("/config")
async def get_config():
    return {
        "arc_chain_id": ARC_CHAIN_ID,
        "arc_rpc_url": ARC_RPC_URL,
        "arc_explorer_url": ARC_EXPLORER_URL,
        "escrow_address": ESCROW_ADDRESS,
        "nanopayment_usdc": NANOPAYMENT_USDC,
    }


if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="127.0.0.1", port=ORCHESTRATOR_PORT)

File: agent-ts/packages/services/src/orchestrator.ts
/**
 * Orchestrator service (PRD S6.2): decomposes a job, pays specialists via
 * x402, hands the assembled claims to the (deterministic) evaluator, then
 * triggers settlement. Owns the JobRecord end to end since it's the only
 * role that needs to persist job state across the whole lifecycle. Ported
 * from agent/agents/orchestrator.py.
 *
 * Run standalone with:
 *   npm run orchestrator --workspace=@arcproof/services
 * This is also the one HTTP surface a frontend integrates against: POST
 * /jobs to submit work, GET /jobs/:id to poll status, GET /jobs/:id/logs
 * for live activity while a job is still processing, GET /reputation for
 * the dashboard feed.
 */
import Fastify from "fastify";
import { randomUUID } from "node:crypto";

import {
  JobRequestSchema,
  type JobRecord,
  type Claim,
  config,
  x402,
  wallet,
  chain,
  escrowContract,
  settlement,
  store,
} from "@arcproof/core";
import { planSpecialists, writeMemo } from "./langchainPlanner.js";
import { registerSecurity, checkApiKey } from "./security.js";
import { startJobLog, logEvent, getJobLog } from "./jobLog.js";

const app = Fastify({ logger: false });
await registerSecurity(app, { cors: true });

const SPECIALIST_URLS: Record<string, string> = {
  "onchain-agent-v1": config.ONCHAIN_AGENT_URL,
  "news-agent-v1": config.NEWS_AGENT_URL,
  "compliance-agent-v1": config.COMPLIANCE_AGENT_URL,
};

const ORCHESTRATOR_PRIVATE_KEY = config.WALLETS.orchestrator.privateKey;

function explorerTxUrl(txHash: string): string {
  return `${config.ARC_EXPLORER_URL}/tx/${txHash}`;
}

interface SpecialistCallResult {
  claims: Claim[];
  nanopaymentTxHash: string | null;
}

async function callSpecialist(
  name: string,
  jobReq: { protocol_slug: string; target_address?: string | null; inject_fault?: string | null },
  jobId: string
): Promise<SpecialistCallResult> {
  const url = SPECIALIST_URLS[name];
  console.log(`[orchestrator] -> calling ${name} (${url}/analyze)`);
  const { response, txHash } = await x402.x402Post(`${url}/analyze`, ORCHESTRATOR_PRIVATE_KEY, {
    job_id: jobId,
    protocol_slug: jobReq.protocol_slug,
    target_address: jobReq.target_address,
    inject_fault: jobReq.inject_fault,
  });
  if (!response.ok) throw new Error(`specialist ${name} returned ${response.status}: ${await response.text()}`);
  if (txHash) {
    logEvent(
      jobId,
      "info",
      `Paid ${name} a nanopayment for responding (${config.NANOPAYMENT_USDC} USDC)`,
      { txHash, explorerUrl: explorerTxUrl(txHash) },
      { from: "orchestrator", to: name, kind: "payment" }
    );
  }
  const body = (await response.json()) as { claims: Claim[] };
  return { claims: body.claims, nanopaymentTxHash: txHash };
}

app.post("/jobs", async (request, reply) => {
  if (!checkApiKey(request, reply)) return;
  const jobReq = JobRequestSchema.parse(request.body);
  // A connected browser wallet that already called VeriFiEscrow.lock()
  // itself supplies the same job_id it locked under; otherwise generate a
  // fresh one and lock it ourselves (the CLI / no-wallet-connected path).
  const jobId = jobReq.job_id || `job_${randomUUID().replace(/-/g, "").slice(0, 10)}`;
  startJobLog(jobId);
  console.log(`\n=== NEW JOB ${jobId} -- ${jobReq.protocol_slug} ===`);
  console.log(`[orchestrator] received request: "${jobReq.request_text}" | budget=${jobReq.budget_usdc} USDC | template=${jobReq.template}`);
  logEvent(jobId, "info", `Job received: "${jobReq.request_text}"`, undefined, { from: "requester", to: "orchestrator", kind: "system" });

  const job: JobRecord = {
    job_id: jobId,
    requester_id: jobReq.requester_wallet,
    template: jobReq.template || "unclassified", // replaced with the LLM's inferred label below once planSpecialists() runs
    request_text: jobReq.request_text,
    protocol_slug: jobReq.protocol_slug,
    budget_usdc: jobReq.budget_usdc,
    status: "in_progress",
    created_at: new Date().toISOString(),
    subtasks: [],
    total_paid_usdc: 0,
    claims: [],
    payouts: [],
  };

  if (jobReq.payment_tx_hash) {
    // Requester already locked for real via their own connected wallet --
    // don't trust the claim, independently re-derive it from the contract
    // itself (getJob is a plain view call, same "verify, don't trust"
    // principle x402 already applies to specialist nanopayments).
    let onChain;
    try {
      onChain = await escrowContract.getJob(jobId);
    } catch (e) {
      job.status = "failed";
      await store.jobStore.save(job);
      logEvent(jobId, "error", `Could not read escrow contract state: ${e}`);
      return reply.code(402).send({ detail: `could not read escrow contract state for job ${jobId}: ${e}` });
    }
    const requesterMatches = onChain.requester.toLowerCase() === jobReq.requester_wallet.toLowerCase();
    const lockedEnough = onChain.lockedUsdc + 1e-9 >= jobReq.budget_usdc;
    if (onChain.status !== "locked" || !requesterMatches || !lockedEnough) {
      job.status = "failed";
      await store.jobStore.save(job);
      logEvent(jobId, "error", `Budget lock could not be independently verified on-chain`);
      return reply.code(402).send({
        detail: `job ${jobId} is not verifiably locked on-chain for ${jobReq.budget_usdc} USDC from ${jobReq.requester_wallet} (contract says: status=${onChain.status}, requester=${onChain.requester}, locked=${onChain.lockedUsdc})`,
      });
    }
    job.lock_tx_hash = jobReq.payment_tx_hash;
    console.log(`[orchestrator] budget verified on-chain: ${jobReq.requester_wallet.slice(0, 10)}.. locked ${onChain.lockedUsdc.toFixed(6)} USDC (tx ${jobReq.payment_tx_hash})`);
    logEvent(
      jobId,
      "success",
      `Budget lock independently verified on-chain (${onChain.lockedUsdc.toFixed(4)} USDC)`,
      { txHash: jobReq.payment_tx_hash, explorerUrl: explorerTxUrl(jobReq.payment_tx_hash) },
      { from: "requester", to: "escrow", kind: "payment" }
    );
  } else {
    try {
      const lockTx = await escrowContract.lock(jobId, jobReq.requester_wallet as config.Role, jobReq.budget_usdc);
      job.lock_tx_hash = lockTx.txHash;
      logEvent(
        jobId,
        "success",
        `Locked ${jobReq.budget_usdc} USDC in escrow`,
        { txHash: lockTx.txHash, explorerUrl: lockTx.explorerUrl },
        { from: "requester", to: "escrow", kind: "payment" }
      );
    } catch (e) {
      job.status = "failed";
      await store.jobStore.save(job);
      logEvent(jobId, "error", `Could not lock budget in escrow contract: ${e}`);
      return reply.code(402).send({ detail: `could not lock budget in escrow contract: ${e}` });
    }
  }

  try {
    logEvent(jobId, "info", `Orchestrator's LLM planner is choosing which specialists to engage...`);
    const plan = await planSpecialists(jobReq);
    job.subtasks = plan.specialist_ids;
    if (!jobReq.template) job.template = plan.template_label;
    logEvent(jobId, "info", `Plan: ${plan.specialist_ids.join(", ")} -- ${plan.reasoning}`);

    const allClaims: Claim[] = [];
    const nanopaymentTxByProvider: Record<string, string> = {};
    for (const name of plan.specialist_ids) {
      logEvent(jobId, "info", `Calling ${name}...`, undefined, { from: "orchestrator", to: name, kind: "call" });
      try {
        const { claims, nanopaymentTxHash } = await callSpecialist(name, jobReq, jobId);
        allClaims.push(...claims);
        if (nanopaymentTxHash) nanopaymentTxByProvider[name] = nanopaymentTxHash;
        logEvent(
          jobId,
          "success",
          `${name} responded with ${claims.length} claim${claims.length === 1 ? "" : "s"}`,
          undefined,
          { from: name, to: "orchestrator", kind: "response" }
        );
      } catch (e) {
        console.log(`[orchestrator]   ! ${name} failed: ${e}`);
        logEvent(jobId, "warn", `${name} failed to respond: ${e}`, undefined, { from: name, to: "orchestrator", kind: "response" });
      }
    }

    job.claims = allClaims;
    logEvent(jobId, "info", `Writing the diligence memo...`);
    job.final_memo = await writeMemo(jobReq, allClaims);

    console.log(`[orchestrator] assembled ${allClaims.length} claims from ${plan.specialist_ids.length} specialists -> handing off to evaluator`);
    logEvent(
      jobId,
      "info",
      `Assembled ${allClaims.length} claim${allClaims.length === 1 ? "" : "s"} -- handing off to the evaluator for independent verification`,
      undefined,
      { from: "orchestrator", to: "evaluator-v1", kind: "call" }
    );
    const evalResp = await fetch(`${config.EVALUATOR_URL}/evaluate`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({
        job_id: jobId,
        protocol_slug: jobReq.protocol_slug,
        target_address: jobReq.target_address,
        claims: allClaims,
      }),
    });
    if (!evalResp.ok) throw new Error(`evaluator returned ${evalResp.status}: ${await evalResp.text()}`);
    const evalBody = (await evalResp.json()) as { claims: Claim[] };
    job.claims = evalBody.claims;

    for (const c of job.claims) {
      const level = c.verification_status === "match" ? "success" : c.verification_status === "mismatch" ? "error" : "warn";
      const icon = c.verification_status === "match" ? "MATCH" : c.verification_status === "mismatch" ? "MISMATCH" : "UNVERIFIABLE";
      logEvent(
        jobId,
        level,
        `[${c.claim_type}] ${icon} -- ${c.claim_text}${c.verification_note ? ` (${c.verification_note})` : ""}`,
        undefined,
        { from: "evaluator-v1", to: c.provider_agent_id, kind: "verdict" }
      );
    }

    if (!settlement.hasCheckableClaims(job.claims)) {
      // Every specialist failed, or every claim came back unverifiable --
      // computeJobVerdict would call this a clean "accept" (0 mismatches),
      // which finalize()s the contract with the full budget withheld
      // forever instead of refunded. Throw so the catch block below runs
      // the existing refund path instead of settling.
      throw new Error(`no checkable claims for job ${jobId} (every specialist failed or all claims were unverifiable) -- refunding rather than auto-accepting`);
    }

    logEvent(jobId, "info", `Settling per-specialist payouts on-chain...`);
    const settled = await settlement.settle(job, (event) => {
      const actors =
        event.type === "release"
          ? { from: "escrow", to: event.providerId, kind: "settlement" as const }
          : { from: "orchestrator", to: "escrow", kind: "system" as const };
      logEvent(jobId, "success", event.message, { txHash: event.txHash, explorerUrl: event.explorerUrl }, actors);
    });

    // Fold each specialist's earlier nanopayment tx hash into its final
    // payout record now that settle() has produced the payouts array.
    for (const payout of settled.payouts) {
      const tx = nanopaymentTxByProvider[payout.provider_agent_id];
      if (tx) payout.nanopayment_tx_hash = tx;
    }

    await store.jobStore.save(settled);

    console.log(`[orchestrator] job ${jobId} DONE -- verdict=${settled.overall_verdict?.toUpperCase()} paid=${settled.total_paid_usdc.toFixed(4)} USDC`);
    logEvent(jobId, "success", `Job complete -- verdict ${settled.overall_verdict?.toUpperCase()}, ${settled.total_paid_usdc.toFixed(4)} USDC paid`);
    return settled;
  } catch (e) {
    // Budget is already locked in the escrow contract at this point -- any
    // failure past that line must not leave real funds stranded with no
    // job record and no way back to the requester. refund() is a real
    // contract call, same trust model as any release.
    console.log(`[orchestrator] job ${jobId} failed after budget lock: ${e} -- refunding escrow contract -> requester`);
    logEvent(jobId, "warn", `Job failed after budget lock: ${e} -- refunding escrow contract`);
    job.status = "failed";
    try {
      const refundTx = await escrowContract.refund(jobId);
      job.refund_tx_hash = refundTx.txHash;
      logEvent(
        jobId,
        "success",
        `Refunded full locked budget back to the requester`,
        { txHash: refundTx.txHash, explorerUrl: refundTx.explorerUrl },
        { from: "escrow", to: "requester", kind: "payment" }
      );
    } catch (refundError) {
      console.log(`[orchestrator]   ! refund ALSO failed: ${refundError} -- funds remain locked in contract, job marked failed`);
      logEvent(jobId, "error", `Refund ALSO failed: ${refundError} -- funds remain locked in the contract`);
    }
    await store.jobStore.save(job);
    return reply.code(500).send({ detail: `job failed after budget was locked; escrow contract refunded to requester where possible: ${e}` });
  }
});

app.get("/jobs/:jobId", async (request, reply) => {
  const { jobId } = request.params as { jobId: string };
  const job = store.jobStore.get(jobId);
  if (!job) return reply.code(404).send({ detail: "job not found" });
  return job;
});

app.get("/jobs/:jobId/logs", async (request) => {
  const { jobId } = request.params as { jobId: string };
  return { logs: getJobLog(jobId) };
});

app.get("/jobs", async () => store.jobStore.listAll());

app.get("/reputation", async () => store.reputationStore.listAll());

app.get("/wallets", async () => {
  const plain = await wallet.ledger.allBalances();
  const balances: Record<string, number> = {};
  for (const [role, bal] of Object.entries(plain)) {
    const circleEntry = config.CIRCLE_WALLETS[role as config.Role];
    if (circleEntry) {
      // This role signs its contract calls through the Circle wallet now
      // (see escrowContract.ts) -- that's the balance that actually moves.
      balances[`${role}-circle`] = await chain.getBalanceUsdc(circleEntry.address);
    } else {
      balances[role] = bal;
    }
  }
  try {
    balances["escrow-contract"] = await chain.getBalanceUsdc(escrowContract.contractAddress());
  } catch {
    /* contract not deployed yet -- omit rather than fail the whole endpoint */
  }
  return balances;
});

app.get("/config", async () => ({
  arc_chain_id: config.ARC_CHAIN_ID,
  arc_rpc_url: config.ARC_RPC_URL,
  arc_explorer_url: config.ARC_EXPLORER_URL,
  escrow_contract_address: escrowContract.contractAddress(),
  nanopayment_usdc: config.NANOPAYMENT_USDC,
}));

export { app };

if (import.meta.url === `file://${process.argv[1]}`) {
  app.listen({ port: config.ORCHESTRATOR_PORT, host: "127.0.0.1" }).then(() => {
    console.log(`[orchestrator] listening on http://127.0.0.1:${config.ORCHESTRATOR_PORT}`);
  });
}
